### PR TITLE
IA32 port fixes

### PR DESCRIPTION
**Pull request by Maxez**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/ia32/_init.S [hunk 1]
```diff
@@ -23,15 +23,14 @@
 .section .init
 
 
-/* Loader high level part entry point */
+/* External variables */
+.extern _plo_size
```
Comment (gerard5) [inline hal/ia32/_init.S:27]: Add:
```suggestion
.extern _plo_size
.extern _plo_seg16
```

File: hal/ia32/_init.S [hunk 2]
```diff
@@ -41,115 +40,86 @@ _start:
...
-	movl %eax, _plo_geo_secs
+	/* Set start CHS address */
+	xorw %bp, %bp
+	xorb %dh, %dh
+	movw $0x1, %ax
 
 _plo_load:
 	/* Load high level part into memory */
-	movl $_plo_size-1, %ecx
+	movw $((ADDR_PLO + 0x200) >> 4), %bx
```
Comment (gerard5) [inline hal/ia32/_init.S:94]: Replace with:
```suggestion
	movw $(_plo_seg16), %bx
```

**Review by gerard5 — COMMENTED** [review]
@Maxez I like the changes, `.init` (boot sector, stage0) is much lighter now. This `$((ADDR_PLO + 0x200) >> 4)` still looks weird-strange to me, and we can take the 16bit segment from the linker script, we agreed to resign from using macrodefs as much as possible in the asm, so this offset calculation can be replaced with the linker symbol which delivers the computed 16-bit segment at which the plo `.text` (0x7e0) is loaded, and it also gives us a better way to automatically calculate the segment based on the location of `.text` we know it in the linker script, so segment can be easliy computed without constant offsets.

See my comment's suggestions (add/replace), it works pretty well after the following patch is applied. I've decided to post patch because I can't comment no existing changes to the linker script (lds) in the PR until now.

```diff
diff --git a/hal/ia32/_init.S b/hal/ia32/_init.S
index 9a35cc3..4f77165 100644
--- a/hal/ia32/_init.S
+++ b/hal/ia32/_init.S
@@ -25,6 +25,7 @@
 
 /* External variables */
 .extern _plo_size
+.extern _plo_seg16
 .extern _stack
 .extern _stacksz
 .extern _startc
@@ -66,7 +67,7 @@ _plo_start:
 
 _plo_load:
 `       /* Load high level part into memory */
-`       movw $((ADDR_PLO + 0x200) >> 4), %bx
+`       movw $(_plo_seg16), %bx
 `       movw %bx, %es
 `       xorw %bx, %bx
 `       xorb %dh, %dh
diff --git a/ld/common/plo-ia32.lds b/ld/common/plo-ia32.lds
index 992e4d8..85223fb 100644
--- a/ld/common/plo-ia32.lds
+++ b/ld/common/plo-ia32.lds
@@ -133,5 +133,8 @@ SECTIONS
 `       _plo_size = LOADADDR(.data) + SIZEOF(.data) - ORIGIN(PLO_IMAGE);
 `       _plo_load_addr = ORIGIN(PLO_IMAGE);
 
+`       /* 16bit real-mode segments */
+`       _plo_seg16 = LOADADDR(.text) >> 4;
+
 `       /DISCARD/ : { *(.note.GNU-stack) *(.gnu_debuglink) *(.gnu.lto_*) }
 }
diff --git a/ld/ia32-generic.ldt b/ld/ia32-generic.ldt
index 6e24394..c7750c1 100644
--- a/ld/ia32-generic.ldt
+++ b/ld/ia32-generic.ldt
@@ -33,8 +33,6 @@
 #define ADDR_PDIR   0x3000
 #define ADDR_PTABLE 0x4000
 
-#define ADDR_PLO 0x7c00
-
 /* Disk caches (below upper memory starting at 0x80000) */
 #define ADDR_RCACHE 0x78000
 #define SIZE_RCACHE 0x4000
```

**Review by agkaminski — COMMENTED** [review]
I see only 2 small misra related nitpicks

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**